sympy/polys/numberfields/modules.py

```diff
@@ -0,0 +1,1242 @@
+        ``NotImplemented``.
+        """
+        if self.is_compat(other):
+            d, e = self.denom, other.denom
+            m = ilcm(d, e)
+            u, v = m // d, m // e
+            col = to_col([u * a + v * b for a, b in zip(self.coeffs, other.coeffs)])
+            return type(self)(self.module, col, denom=m).reduced()
+        elif isinstance(other, ModuleElement):
+            a, b = self.make_compat(other)
+            if a is None:
+                return NotImplemented
+            return a + b
+        elif is_rat(other):
+            return self + self.module.element_from_rational(other)
+        return NotImplemented
+
+    __radd__ = __add__
+
+    def __neg__(self):
+        return self * -1
+
+    def __sub__(self, other):
+        return self + (-other)
+
+    def __rsub__(self, other):
+        return -self + other
+
+    def __mul__(self, other):
+        """
+        A ModuleElement can be multiplied by a rational number, or by another
+        ModuleElement.
+
+        When the multiplier is a rational number, the product is computed by
+        operating directly on the coefficients of this ModuleElement.
+
+        When the multiplier is another ModuleElement, the product will belong
+        to the nearest common ancestor (NCA) of the modules of the two operands,
+        and that NCA must have a multiplication table. If the NCA does not
+        exist, we return ``NotImplemented``. If the NCA does not have a mult.
+        table, ``ClosureFailure`` will be raised.
+        """
+        if self.is_compat(other):
+            M = self.module.mult_tab()
+            A, B = self.col.flat(), other.col.flat()
+            n = self.n
+            C = [0] * n
+            for u in range(n):
+                for v in range(u, n):
+                    c = A[u] * B[v]
+                    if v > u:
+                        c += A[v] * B[u]
+                    if c != 0:
+                        R = M[u][v]
+                        for k in range(n):
+                            C[k] += c * R[k]
+            d = self.denom * other.denom
+            return self.from_int_list(self.module, C, denom=d)
+        elif isinstance(other, ModuleElement):
+            a, b = self.make_compat(other)
+            if a is None:
+                return NotImplemented
+            return a * b
+        elif is_rat(other):
+            a, b = get_num_denom(other)
+            if a == b == 1:
+                return self.copy()
+            else:
+                return make_mod_elt(self.module,
+                                 self.col * a, denom=self.denom * b).reduced()
+        return NotImplemented
+
+    __rmul__ = __mul__
+
+    def _zeroth_power(self):
+        return self.module.one()
+
+    def _first_power(self):
+        return self.copy()
+
+    def __floordiv__(self, a):
+        if is_rat(a):
+            a = QQ(a)
+            if a == 0:
+                raise ZeroDivisionError
+            return self * (1/a)
```

Comment:
We don't need to explicitly raise ZeroDivisionError here because `1/QQ(0)` already does that.